### PR TITLE
MsSql supports for Window Function

### PR DESCRIPTION
**Pull request by rohitgaurDatametica**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java [hunk 1]
```diff
@@ -114,8 +115,25 @@ public MssqlSqlDialect(Context context) {
       case TRIM:
         unparseTrim(writer, call, leftPrec, rightPrec);
         break;
+      case CONCAT:
+        final SqlWriter.Frame concatFrame = writer.startFunCall("CONCAT");
+        for (SqlNode operand : call.getOperandList()) {
+          writer.sep(",");
```
Comment (roy-sounak) [inline core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java:121]: Is there a better way to implement the ','

File: core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java [hunk 2]
```diff
@@ -308,6 +334,41 @@ private void unparseTrim(SqlWriter writer, SqlCall call, int leftPrec, int right
...
+  private void unparseSqlWindow(SqlWindow sqlWindow, SqlWriter writer, SqlCall call) {
+    final SqlWindow window = sqlWindow;
+    writer.print("OVER ");
+    SqlCall operand1 = call.operand(0);
+    final SqlWriter.Frame frame =
+            writer.startList(SqlWriter.FrameTypeEnum.WINDOW, "(", ")");
+    if (window.getRefName() != null) {
+      window.getRefName().unparse(writer, 0, 0);
+    }
+    if (((SqlBasicCall) call.operand(0)).getOperandList().isEmpty()
```
Comment (roy-sounak) [inline core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java:350]: Please remove coagulated if-else blocks.

File: core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java [hunk 3]
```diff
@@ -308,6 +334,41 @@ private void unparseTrim(SqlWriter writer, SqlCall call, int leftPrec, int right
...
+        writer.sep("PARTITION BY");
+        final SqlWriter.Frame partitionFrame = writer.startList("", "");
+        window.getPartitionList().unparse(writer, 0, 0);
+        writer.endList(partitionFrame);
+      }
+      writer.print("ORDER BY ");
+      SqlNode operand2 = operand1.operand(0);
+      operand2.unparse(writer, 0, 0);
+
+      writer.print("ROWS BETWEEN ");
```
Comment (roy-sounak) [inline core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java:365]: Please reduce function calls. Please use sep() for separators and print() to write strings

File: core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java [hunk 4]
```diff
@@ -308,6 +334,41 @@ private void unparseTrim(SqlWriter writer, SqlCall call, int leftPrec, int right
...
+      writer.endList(frame);
+    } else if (window.getOrderList().size() == 0) {
+      if (window.getPartitionList().size() > 0) {
+        writer.sep("PARTITION BY");
+        final SqlWriter.Frame partitionFrame = writer.startList("", "");
+        window.getPartitionList().unparse(writer, 0, 0);
+        writer.endList(partitionFrame);
+      }
+      writer.print("ORDER BY ");
+      SqlNode operand2 = operand1.operand(0);
```
Comment (roy-sounak) [inline core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java:362]: Please change var names Operand1 and operand2. Use descriptive names.

**Review by roy-sounak — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 5]
```diff
@@ -6504,9 +6531,13 @@ public void testWindowFunctionWithOrderByWithoutcolumn() {
...
             + "PRECEDING AND UNBOUNDED FOLLOWING)\n"
             + "FROM \"foodmart\".\"employee\"";
+    final String synapseSql = "SELECT COUNT(*) OVER ()\n"
+            + "FROM [foodmart].[employee]";
     sql(query)
             .withSnowflake()
-            .ok(expectedSnowflake);
+            .ok(expectedSnowflake)
+            .withMssql()
+            .ok(synapseSql);
```
Comment (roy-sounak) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:6540]: Please add test cases for dateofthemonth and concat

**Review by roy-sounak — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java (hunk 1 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java:121]: writing in last will add an extra comma in the end

File: core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java (hunk 2 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java]: done

**Review by roy-sounak — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 6]
```diff
@@ -6635,6 +6673,19 @@ public void testCaseForLnFunction() {
       .withMssql()
       .ok(expected);
   }
+
+  @Test public void testForDayOfMonth() {
+    final RelBuilder builder = relBuilder().scan("EMP");
+    final RexNode condition = builder.call(SqlStdOperatorTable.DAYOFMONTH,
+            builder.field("HIREDATE"));
```
Comment (soma-mondal) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:6680]: Please do not add test cases here. add in SqlFunctions

**Review by soma-mondal — APPROVED** [review]

**Review by deepakDatametica — APPROVED** [review]

**Pull request merged**